### PR TITLE
(Bugfix) Finalize claim UX

### PR DESCRIPTION
**Pull request by L03TJ3**
# Description
List of issues can be found issue 297

About # (link your issue here)
https://github.com/GoodDollar/GoodProtocolUI/issues/297

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

# Checklist:
- [x] PR title matches follow: (Feature|Bug|Chore) Task Name
- [x] My code follows the style guidelines of this project
- [x] I have followed all the instructions described in the initial task (check Definitions of Done)
- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have added reference to a related issue in the repository
- [x] I have added a detailed description of the changes proposed in the pull request. I am as descriptive as possible, assisting reviewers as much as possible.
- [ ] I have added screenshots related to my pull request (for frontend tasks)
- [ ] I have pasted a gif showing the feature.
- [x] @mentions of the person or team responsible for reviewing proposed changes

### CODE REVIEW

**Comment by johnsmith-gooddollar** [comment]
@L03TJ3 please link the issue this is related for

File: packages/good-design/src/hooks/useFVModalAction.ts [hunk 1]
```diff
@@ -43,8 +43,8 @@ export const useFVModalAction = ({ firstName, method, onClose }: FVModalActionPr
 
     switch (method) {
       case "redirect": {
-        const link = fvlink?.getLink(firstName, document.location.href, false);
-        console.log("test for redirect back link -->", { link });
+        const callbackUrl = document.location.href.replace("#", "%23");
```
Comment (johnsmith-gooddollar) [inline packages/good-design/src/hooks/useFVModalAction.ts:46]: 1.  fvlink?.getLink should take care of this
2. there's an URL class allowing you to parse / assemble urls and working with all parts correctly including the hash

so you could
a) create URL from callbackUrl
b) add query params using .searchParams.append
c) assemble url back to string. the original hash value should be preserved

**Review by johnsmith-gooddollar — CHANGES_REQUESTED** [review]

File: packages/good-design/src/hooks/useFVModalAction.ts (hunk 1 above)
Comment (johnsmith-gooddollar) [reply packages/good-design/src/hooks/useFVModalAction.ts:46]: @L03TJ3 the issue is at `fvid` side. I'll hot fix gooddapp repo, this could be reverted

**Comment by johnsmith-gooddollar** [comment]
@L03TJ3 

1. url issue fixed here: https://github.com/GoodDollar/GoodDAPP/commit/4dcae7c4f5d8d13226b8e0526ff83cc0d44a94e2
make sure `goodid` DEV & QA redeployed successfully
2. why do you need additional `verifying` flag ? why you pass it onto onClose ? We're sending callbacks / redirecting just after it so it means user have to be verified / failed to verify at this point but not in progress ?
3. fix conflicts please

File: packages/good-design/src/hooks/useFVModalAction.ts [hunk 2]
```diff
@@ -21,9 +21,12 @@ export const useFVModalAction = ({ firstName, method, redirectUrl = defaultRedir
       await fvlink?.getLoginSig();
       await fvlink?.getFvSig();
     } catch (e: any) {
-      setLoading(false);
       return;
+    } finally {
+      setLoading(false);
     }
+    
+    onClose();
```
Comment (L03TJ3) [inline packages/good-design/src/hooks/useFVModalAction.ts:29]: why do we need to close modal specifically?

Comment (johnsmith-gooddollar) [reply packages/good-design/src/hooks/useFVModalAction.ts:29]: then remove it from props, totally, and from claim button which passes onClose callback

Comment (johnsmith-gooddollar) [reply packages/good-design/src/hooks/useFVModalAction.ts:29]: but you removed it just partially I thought it's mistake

File: packages/good-design/src/core/buttons/ClaimButton.tsx [hunk 3]
```diff
@@ -60,69 +114,78 @@ const ClaimButton = ({
...
+          setClaimLoading(true);
+          await handleClaim(true);
+        } else {
+          //// what to do here? tjis should not continue with FV Flow
+        }
+      } else {
+        // here nothing happens, FV flow starts
+        // what to do with an edge case where fuseWhitelisted or syncStatus might not be set or
+        // give a value in time before user interaction is done
       }
```
Comment (johnsmith-gooddollar) [inline packages/good-design/src/core/buttons/ClaimButton.tsx:163]: just simplify this by now
```suggestion
      if (isWhitelisted) {
        setClaimLoading(true);
        await handleClaim(first);
        return;
      } 
      
      if (fuseWhitelisted && syncStatus) {
        const success = await syncStatus;
        
        if (!success) {
          return;
        }
        
        setClaimLoading(true);
        await handleClaim(true);
      }          
```

File: packages/good-design/src/hooks/useFVModalAction.ts [hunk 4]
```diff
@@ -9,22 +9,24 @@ interface FVModalActionProps extends Pick<FVFlowProps, "method" | "firstName"> {
...
 
   const verify = useCallback(async () => {
     setLoading(true);
 
-    await fvlink?.getLoginSig();
-    await fvlink?.getFvSig();
+    try {
+      await fvlink?.getLoginSig();
+      await fvlink?.getFvSig();
+    } catch (e: any) {
```
Comment (johnsmith-gooddollar) [inline packages/good-design/src/hooks/useFVModalAction.ts:23]:
```suggestion
    } catch {
```

File: packages/sdk-v2/src/sdk/base/sdk.ts [hunk 5]
```diff
@@ -76,11 +76,13 @@ export class BaseSDK {
       signer
         .getAddress()
         .then(async () => void (this.signer = signer))
-        .catch(e => {
-          console.warn("BaseSDK: provider has no signer", { signer, provider, e });
+        .catch(() => {
+          // todo @l03tj3: revert back
```
Comment (johnsmith-gooddollar) [inline packages/sdk-v2/src/sdk/base/sdk.ts:80]: if the PR isn't a draft anymore, it shouldn't contain any TODOs (only if they're out of this PR's scope and you're going to open separate tasks in the icebox)

**Review by johnsmith-gooddollar — CHANGES_REQUESTED** [review]

File: packages/sdk-v2/src/sdk/claim/react.ts [hunk 6]
```diff
@@ -46,6 +43,7 @@ export const useClaim = (refresh: QueryParams["refresh"] = "never") => {
   const identity = useGetContract("Identity", true, "claim", chainId) as IIdentity;
   const claimCall = useContractFunction(ubi, "claim");
 
+  console.log("useClaim -->", { account, chainId, refreshOrNever });
```
[the remark this answers is not in the document]
Comment (L03TJ3) [reply packages/sdk-v2/src/sdk/claim/react.ts]: done

File: packages/sdk-v2/src/sdk/base/sdk.ts (hunk 5 above)
Comment (L03TJ3) [reply packages/sdk-v2/src/sdk/base/sdk.ts]: done

File: packages/good-design/src/hooks/useFVModalAction.ts (hunk 4 above)
Comment (L03TJ3) [reply packages/good-design/src/hooks/useFVModalAction.ts]: done

File: packages/good-design/src/advanced/web3action/Web3Action.tsx [hunk 7]
```diff
@@ -73,7 +73,7 @@ export const Web3ActionButton: FC<Web3ActionProps> = withTheme({ name: "Web3Acti
       const connectFn = handleConnect || (activateBrowserWallet as any);
       const isConnected = await connectFn().catch(throwIfCancelled);
 
-      if (handleConnect && !isConnected) {
+      if (handleConnect && !isConnected.length) {
```
[the remark this answers is not in the document]
Comment (L03TJ3) [reply packages/good-design/src/advanced/web3action/Web3Action.tsx]: was the last part, now the flow expects boolean as return from connectFn

File: packages/good-design/src/core/buttons/ClaimButton.tsx (hunk 3 above)
Comment (L03TJ3) [reply packages/good-design/src/core/buttons/ClaimButton.tsx]: removed for now

File: packages/good-design/src/core/buttons/types.ts [hunk 8]
```diff
@@ -4,11 +4,12 @@ export interface FVFlowProps {
   firstName: string;
   method: "popup" | "redirect";
   redirectUrl?: string;
+  claim: () => Promise<boolean>;
   chainId?: number;
   styles?: any;
   claimed?: boolean;
   refresh?: "everyBlock" | "never" | number | undefined;
-  claim: () => Promise<boolean>;
+  handleConnect?: () => Promise<any> | void;
```
[the remark this answers is not in the document]
Comment (L03TJ3) [reply packages/good-design/src/core/buttons/types.ts]: done

**Review by johnsmith-gooddollar — APPROVED** [review]

**Pull request merged**